service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/disk/DetachAzureDiskFromVmStep.java

```diff
@@ -0,0 +1,190 @@
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    final AzureCloudContext azureCloudContext =
+        DeleteAzureDiskFlightUtils.getAzureCloudContext(context);
+    final String diskResourceId =
+        DeleteAzureDiskFlightUtils.getAzureDiskResourceId(azureCloudContext, resource);
+    ComputeManager computeManager = crlService.getComputeManager(azureCloudContext, azureConfig);
+    try {
+      var disk = computeManager.disks().getById(diskResourceId);
+      if (!disk.isAttachedToVirtualMachine()) {
+        logger.info(
+            "Disk {} in workspace {} is not attached to vm.",
+            resource.getResourceId(),
+            resource.getWorkspaceId());
+        return StepResult.getStepResultSuccess();
+      }
+      var vm = getVirtualMachine(computeManager, disk.virtualMachineId());
+      if (vm.isEmpty()) {
+        // this is unlikely event - disk is attached, but the vm cannot be found; let's retry
+        logger.warn(
+            "Disk {} is attached to vm in workspace {}, but vm cannot be found. Retrying.",
+            resource.getResourceId(),
+            resource.getWorkspaceId());
+        return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY);
+      }
+      var attachedDisk =
+          vm.get().dataDisks().entrySet().stream()
+              .filter(entry -> entry.getValue().name().equals(resource.getDiskName()))
+              .findFirst();
+      if (attachedDisk.isPresent()) {
+        detachDisk(vm.get(), attachedDisk.get().getKey());
+      } else {
+        // this is unlikely event - disk is attached, but the vm doesn't
+        // have it in the list of attached disks; let's retry
+        logger.warn(
+            "Disk {} is attached, but not found in the list of attached disk for vm {}.",
+            resource.getResourceId(),
+            disk.virtualMachineId());
+        return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY);
+      }
+    } catch (ManagementException ex) {
+      if (AzureManagementExceptionUtils.isExceptionCode(
+          ex, AzureManagementExceptionUtils.RESOURCE_NOT_FOUND)) {
+        logger.info(
+            "Disk '{}' does not exist in workspace {}.",
+            resource.getResourceId(),
+            resource.getWorkspaceId());
+        return StepResult.getStepResultSuccess();
+      }
+      if (AzureManagementExceptionUtils.isExceptionCode(
+          ex, AzureManagementExceptionUtils.OPERATION_NOT_ALLOWED)) {
+        logger.error(
+            "Failed attempt to detach disk {} from vm in workspace {}.",
+            resource.getResourceId(),
+            resource.getWorkspaceId());
+        return new StepResult(StepStatus.STEP_RESULT_FAILURE_FATAL, ex);
+      }
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, ex);
+    }
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  protected StepResult deleteResource(FlightContext context) throws InterruptedException {
```

Comment:
I don't think the naming convention of `deleteResource` should prevent using it here.

Most of the error cases in the `doStep` of this class are already handled in `DeleteAzureControlledResourceStep`.

Then if there is additional error handling, you can override the exception handling method for just the cases you care about.

As it's currently written, you might as well use the `DeleteControlledResourceStep`.

Reply:
In this case, I definitely intended to use an interface, but started using the abstract class as if the interface didn't exist.